src/native/fs/fs_napi.cpp

```diff
@@ -213,6 +224,258 @@ struct Stat : public FSWorker
     }
 };
 
+/**
+ * Getxattr is an fs op
+ */
+struct Getxattr : public FSWorker
+{
+    std::string _path;
+    std::vector<KeyValue> _md;
+    Getxattr(const Napi::CallbackInfo& info) 
+        : FSWorker(info)
+        , _md(0)
+    {
+        _path = info[1].As<Napi::String>();
+        Begin(XSTR() << "Getxattr " << DVAL(_path));
+    }
+    virtual void Work()
+    {
+        ssize_t buf_len, key_len, value_len;
+        char *buf, *key, *val;
+
+        #if !defined(_POSIX_C_SOURCE) || defined(_DARWIN_C_SOURCE)
+                buf_len = listxattr(_path.c_str(), NULL, 0, 0);
+        #else
+                buf_len = listxattr(_path.c_str(), NULL, 0);
+        #endif
+
+        if (buf_len <= 0) {
+            if (buf_len == -1) {
+                SetSyscallError();
+            }
+            return;
+        }
+
+        struct stat stat_res;
+        auto r = stat(_path.c_str(), &stat_res);
+        if (r) {
+            SetSyscallError();
+            return;
+        }
+        auto start_ctime = stat_res.st_ctime;
+
+        buf = static_cast<char*>(malloc(buf_len));
+        if (buf == NULL) {
+            SetSyscallError();
+            return;
+        }
+
+        #if !defined(_POSIX_C_SOURCE) || defined(_DARWIN_C_SOURCE)
+            buf_len = listxattr(_path.c_str(), buf, buf_len, 0);
+        #else
+            buf_len = listxattr(_path.c_str(), buf, buf_len);
+        #endif
+
+        if (buf_len == -1) {
+            SetSyscallError();
+            free(buf);
+            return;
+        }
+
+        key = buf;
+        while (buf_len > 0) {
+            // strncmp shouldn't cause error if length of key < USER_MD_PREFIX
```

Comment:
You should construct an std::string for every key right at the start.
Since std::string can be constructed from null terminated c strings, this is as simple as:
```
std::string key(buf);
buf += key.size() + 1;
buf_len -= key.size() + 1;
```